QueryEditor: Don't use _.defaults as it mutates the query prop directly

ashharrison90 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

- when changing datasources from `--Grafana--` to `Influx` (or any other datasource really), the query still contains `queryType: 'randomWalk'`
- this is because `QueryGroup` rerenders before `dsSettings` is set correctly in the state at https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryGroup.tsx#L153. `setState` is asynchronous, no guarantee it will update before the next render, etc etc
- when `QueryGroup` rerenders, it rerenders it's children and eventually rerenders the `--Grafana--` `QueryEditor`
- this applies defaults here: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/grafana/components/QueryEditor.tsx#L248, which looks like it directly mutates the `query` prop. so the very act of rerendering causes `query` to be modified with `queryType: 'randomWalk'`
- instead, let's apply the default on mount in an immutable way

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

For https://github.com/grafana/grafana/issues/37800

**Special notes for your reviewer**:

ashharrison90 committed:
QueryEditor: Don't use _.defaults as it mutates the query prop directly

ashharrison90 committed:
QueryEditor: Update the query type on mount if none exists

ashharrison90 commented:
@torkelo @ryantxu i've done some initial testing and it _seems_ ok, but have added you both directly as i'm not sure whether there's some magic behaviour that relies on the fact that it used to mutate the query prop 😕

ryantxu reviewed (approved):
This is a good change!,  ~~but I'm not sure this really "fixes" #37800.~~

~~When we switch data sources right now, the old query is still sent to the new data source -- sometimes that is fine.~~

ashharrison90 commented:
i've changed it to `For` instead of `Fixes` for now 😄 

@ryantxu is that true? i thought https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryGroup.tsx#L136 would clear the queries when changing datasources, so the only thing that should get sent to the new datasource is `{refId: 'A'}`

ryantxu commented:
> is that true?

um -- look at that!  I take back my comment :)

torkelo commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -65,6 +63,13 @@ export class QueryEditor extends PureComponent<Props, State> {
+    const { query } = this.props;
+
+    // If there is no query type, set the default
+    if (!query.queryType) {
+      const defaultQueryType = this.queryTypes.find((v) => v.value === defaultQuery.queryType) || this.queryTypes[0];
+      this.onQueryTypeChange(defaultQueryType);
+    }
```
We should not call model change callbacks on mount, is there a way to avoid this? (calling this triggers a bunch of events, and model persistence changes that should really be due to user action, not just mount)

ashharrison90 replied:
@torkelo guess it depends if you consider a query with just `{refId: 'A'}` in the json valid? if so, we can just not call anything on mount and it will all work. https://github.com/grafana/grafana/pull/37965/files#diff-1c565d0c1df80b370a946c7d533c0ea5a534593b5db63202e2c15075a4a51465R262 ensures if there's no matching `queryType` we default to the first. and as soon as you change the `queryType`, it's updated via `onQueryTypeChange`. that feels a little weird though because then there's a disconnect between the json model and what's displayed on screen.

torkelo replied:
we need a way to set query defaults, not sure where that should be though.

ashharrison90 replied:
hmmm i see

`QueryGroup` has access to the datasource, so we could put an optional `defaultQuery` in the plugin.json file under `queryOptions` i guess.

i'm not sure what changes would be required to support that in the backend though, or if that even makes sense. i guess you'd then have to include `"refId": "A"` in that default query, which seems redundant. and you wouldn't be able to use any of the defined types. e.g. for the --Grafana-- datasource, there's an [enum for the query types](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/grafana/types.ts#L8) (and a [default query](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/grafana/types.ts#L21)) but if we put that in json we'd just have to know the internal strings and wouldn't be able to use those 😕 

i think i'm talking myself out of that already... 🤔

ashharrison90 committed:
QueryEditor: Mutate in the constructor instead of render

ashharrison90 commented:
@torkelo i've updated the title and added details to https://github.com/grafana/grafana/issues/37800 so we can use that to track implementing a mechanism for providing a default query for each datasource.

i've also updated this to be more of a temporary fix. it's doing the exact same thing as before, but in the constructor instead of the render method. that way subsequent rerenders won't overwrite the query.

torkelo commented on public/app/plugins/datasource/grafana/components/QueryEditor.tsx:
```diff
@@ -245,7 +251,8 @@ export class QueryEditor extends PureComponent<Props, State> {
 
   render() {
-    const query = defaults(this.props.query, defaultQuery);
+    const { query } = this.props;
```
can we not apply the defaults immutably here in render instead of doing it mutably in the constructor? or does the rest of the component rely on the mutation of props.query?

ashharrison90 replied:
@torkelo we can, the only reason i didn't want to do that was that it then introduces a disconnect between the json and what's displayed on screen.

## when switching to the --Grafana-- datasource and not modifying anything

applying it mutably in the constructor:
```json
      ...
      "targets": [
        {
          "queryType": "randomWalk",
          "refId": "A"
        }
      ]
      ...
```

applying it immutably in render:
```json
      ...
      <no targets key at all>
      ...
```

in both cases everything works as expected in the UI. tbh i'm not sure which is worse, the disconnect or the mutating. 😅  happy to defer to your wisdom 🧙

torkelo replied:
defaults should not be saved anyway (ideally), so that is not a big issue

ashharrison90 replied:
cool, that makes things easier. i've updated to apply them immutably in the render method 👍

ashharrison90 committed:
Apply defaults immutably in render
